### ISSUE TITLE
incorrect param update in sempre1.0

### ISSUE THREAD

**Issue by uwittygit** [percyliang/sempre#79]
Hi, Liang
I  met a problem when I tried to update the params with sempre1.0
let x be sentence, let y be the class
(1) original formula: score(x,y) = sum(feature(x,y)) =  w1 + w2 + ..... wn,  W was the params
      feature was added to indicatorFeatures
       (score, W takes effect when calculate gradient)
     with SGD, the model can be trained to an accuracy 88% (small data).
     
(2) new formula: score = sum(feature(x,y)) =  w1 * 0.1 + w2 * 0.1 + ..... wn * 0.1
     feature was added to generalFeatures 
     (score, W and 0.1 takes effect when calculate gradient)
     with a fixed 0.1 factor, the model accuracy should be the same,right?
     but they are not, the new accuracy is 76%, Why?

(3) Ｉmade a different try,  sum(feature(x,y)) =  w1 + w2 + ..... wn,  score = 0.1* sum(feature(x,y))
       W and score was for SGD trainning.
        (score, W takes effect when calculate gradient)
      feature was added to indicatorFeatures
     guess what, the accuracy  returned to 88% again.

Summary: it seems that if I add any factor to SGD trainning process ,the accuracy drops a lot. the accuracy was infected by -Params.l1RegCoeff 0.00562341325. 
I guess the problem was caused by L1 reg, It seems to be designed with fixed factor = 1.

do you have any idea how to deal with this issue?
by the way how do you get -Params.l1RegCoeff 0.00562341325, i want to generate a new one for my model.
Thanks ,hope you can reproduce this issue.

**Comment by ppasupat** [comment]
Hi uwittygit,

Thank you for reporting the issue. Would you mind attaching the dataset you used to produce the result?

**Comment by ppasupat** [comment]
Alternatively, could you try using SEMPRE 2.0 instead? The parameter updates and lazy regularization were cleaned up in the new version.

**Comment by uwittygit** [comment]
I make a lot of modifications to sempre1.0
removed freebase and question generations, a lot works to accelerate parsing process
my dataset looks very different with sempre's. but the FeatureVector and SGD was untouched. 

I think there is an easy way for you to reproduce this issue, 
change code in Aligner.java,  in every place there is "featureVector.add"
such as:
featureVector.add("SyntAlign",rule.toString()); --> featureVector.add("SyntAlign",rule.toString(), 0.1);

**Comment by uwittygit** [comment]
Hi ppasupat,
multiply learning rate (stepSize) by 10 and  divide L1 Penal param (l1RegCoeff) by 10, the Trainning process of (2) is identical to (3). the accuracy recovered.

in the paper "Stochastic Gradient Descent Training for
L1-regularized Log-linear Models with Cumulative Penalty", it says: "where C is the meta-parameter that controls the
degree of regularization, which is usually tuned by
cross-validation or using the heldout data."

how did you tuned this penalty param?

**Comment by ppasupat** [comment]
Glad that the issue is sorted out. We choose the hyperparameter C by grid search (run each value of C on 3 train-dev splits, and choose the C with the highest average accuracy)

**Comment by uwittygit** [comment]
@ppasupat  thanks !
